dagfactory/dagbuilder.py

```diff
@@ -624,6 +610,104 @@ def replace_expand_values(task_conf: Dict, tasks_dict: Dict[str, BaseOperator]):
                     task_conf["expand"][expand_key] = tasks_dict[task_id].output
         return task_conf
 
+    @staticmethod
+    def evaluate_condition_with_datasets(condition_string: str, datasets_filter: List[str]) -> Any:
+        """
+        Evaluates a condition using the dataset filter, transforming URIs into valid variable names.
+
+        :param condition_string: A string representing the logical condition to evaluate.
+            Example: "(dataset_custom_1 & dataset_custom_2) | dataset_custom_3".
+        :type condition_string: str
+        :param datasets_filter: A list of dataset URIs to be evaluated in the condition.
+        :type datasets_filter: List[str]
+
+        :returns: The result of the logical condition evaluation with URIs replaced by valid variable names.
+        :rtype: Any
+        """
+        dataset_map = {}
+        for uri in datasets_filter:
+            valid_variable_name = utils.make_valid_variable_name(uri)
+            condition_string = condition_string.replace(uri, valid_variable_name)
+            dataset_map[valid_variable_name] = Dataset(uri)
+        evaluated_condition = eval(condition_string, {}, dataset_map)
+        return evaluated_condition
+
+    @staticmethod
+    def process_file_with_datasets(
+        file: str, datasets_filter: List[str], condition_string: Optional[str] = None
+    ) -> Any:
+        """
+        Processes datasets from a file and evaluates conditions if provided.
+
+        :param file: The file path containing dataset information in a YAML or other structured format.
+        :type file: str
+        :param datasets_filter: A list of dataset names to filter and process.
+        :type datasets_filter: List[str]
+        :param condition_string: A logical condition string to evaluate using the datasets.
+            If not provided, the function returns a list of `Dataset` objects based on the file and filter.
+            Example: "(dataset_custom_1 & dataset_custom_2) | dataset_custom_3".
+        :type condition_string: Optional[str]
+
+        :returns: The result of the condition evaluation if `condition_string` is provided, otherwise a list of `Dataset` objects.
+        :rtype: Any
+        """
+        is_airflow_version_at_least_2_9 = version.parse(AIRFLOW_VERSION) >= version.parse("2.9.0")
+        if condition_string and is_airflow_version_at_least_2_9:
+            map_datasets = utils.get_datasets_map_uri_yaml_file(file, datasets_filter)
+            dataset_map = {alias_dataset: Dataset(uri) for alias_dataset, uri in map_datasets.items()}
+            return eval(condition_string, {}, dataset_map)
```

Comment:
Using eval can lead to security risks, debugging, and maintainability challenges.

The most critical drawback of eval is that it executes arbitrary code. DAG Factory users can execute arbitrary code using `BashOperator` on Airflow worker nodes at a task execution level. However, with the proposed change, we're allowing end-users to execute arbitrary code at a scheduler / DAG processor level, so the damage caused can have more significant implications.

What do you think about implementing a function [ast](https://docs.python.org/3/library/ast.html) to resolve this expression? It has very few node types (and, or, parenthesis and the actual literals), so it should not be complex.